Data/ByteString/Builder/RealFloat.hsc

```diff
@@ -0,0 +1,289 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE ForeignFunctionInterface #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE MagicHash, UnboxedTuples #-}
+
+module Data.ByteString.Builder.RealFloat
+  ( FFFormat(..)
+  , floatDec
+  , doubleDec
+  , formatFloat
+  , formatDouble
+  ) where
+
+import Data.ByteString.Internal (ByteString(..), mallocByteString)
+import Data.ByteString.Builder.ASCII (char7, string7)
+import Data.ByteString.Builder.Internal (Builder, byteString)
+
+#if !(MIN_VERSION_base(4,8,0))
+import Data.Monoid
+import Control.Applicative ((<$>))
+#endif
+
+import Foreign.C.Types (CFloat, CDouble, CInt, CUInt, CULong, CUChar)
+
+#if MIN_VERSION_base(4,5,0) || __GLASGOW_HASKELL__ >= 703
+import Foreign.C.Types (CFloat(..), CDouble(..), CInt(..), CUInt(..), CULong(..), CUChar(..))
+#else
+import Foreign.C.Types (CFloat, CDouble, CInt, CUInt, CULong, CUChar)
+#endif
+
+import Foreign.ForeignPtr (ForeignPtr, withForeignPtr)
+import Foreign.Marshal.Alloc (alloca)
+import Foreign.Ptr (Ptr)
+import Foreign.Storable (peek)
+import GHC.Word (Word8, Word32, Word64(..))
+import GHC.Int (Int32)
+import GHC.Float (FFFormat(..), roundTo)
+import GHC.Prim
+import GHC.Show (intToDigit)
+
+#if MIN_VERSION_base(4,4,0)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+#else
+import           GHC.IO (unsafeDupablePerformIO)
+#endif
+
+{-# INLINABLE floatDec #-}
+floatDec :: Float -> Builder
+floatDec = formatFloat FFGeneric Nothing
+
+{-# INLINABLE doubleDec #-}
+doubleDec :: Double -> Builder
+doubleDec = formatDouble FFGeneric Nothing
+
+{-# INLINABLE formatFloat #-}
+formatFloat :: FFFormat -> Maybe Int -> Float -> Builder
+formatFloat fmt prec f =
+  case fmt of
+      FFGeneric
+        | Just b <- specialStr f -> b
```

Comment:
The use of `PatternGuards` requires `default-language:  Haskell2010` in the `.cabal` file, or else an explicit `PatternGuards` extension, but I'd prefer the former. Of course this assumes that @hvr and others are OK with abandoning `Haskell98`.

Reply:
They avoid a nested case but the brevity gain isn't that large so I preemptively changed it. https://github.com/haskell/bytestring/pull/222/commits/12ac021adad2b958c55cc1ec7f9fc11accffd096